### PR TITLE
Add Iterator::exhaust

### LINKED ISSUE #44546: Improve eagerly evaluating an iterator for its side-effects
0b_0101_001_1010 had a good idea [on Reddit](https://www.reddit.com/r/rust/comments/6zq1jn/rust_benchmarking_forloops_map_and_rayon/dmxux5z/). I'll paraphrase it here.

The only way to eagerly consume an iterator, e.g., for its side-effects, on stable Rust is to call `.collect` or `.count` (`.for_each` is nightly only). Most beginners don't associate `.count` with this, so they use `.collect`, which is horrible.

There are a few things we could do to help make it more obvious what the correct thing to do is:

- make `.collect`'s result `#[must_use]`. If you are throwing `.collect`'s result away, you are doing it wrong, and should not have used `.collect` in the first place. A loop or `.for_each` would do.
- make `.count`'s result `#[must_use]`: while `.count` is "efficent" (when compared to `.collect`) it doesn't convey what the code is actually doing
- add an `.eval()` == `.for_each(|_|)`: because `.for_each(|_|)` is also ugly
- address this in docs: `.collect` means copy the results of the iterator into a collection, not "eagerly evaluate", or similar

### PR DESCRIPTION
**Pull request by varkor**
This adds an `Iterator::exhaust` method for evaluating the side effects of an iterator conveniently. There are already ways to do this simply, but a little uglily (e.g. `iter.for_each(|_| ())` or `iter.collect::<()>()`) and it was proposed in #44546 that it'd be nicer to have a dedicated (and discoverable) method to do this.

Closes #44546.

r? @dtolnay

### CODE REVIEW

**Comment by sfackler** [comment]
I think this is what #48945 originally proposed.

**Comment by varkor** [comment]
@sfackler: oh, I hadn't seen that! https://github.com/rust-lang/rust/issues/44546 should probably be closed with a reference to that, then.

File: src/libcore/iter/iterator.rs [hunk 1]
```diff
@@ -2251,6 +2251,32 @@ pub trait Iterator {
...
+    /// // ...
+    ///
+    /// // Trigger the effects!
+    /// count_aloud.exhaust();
+    ///
+    /// ```
+    #[inline]
+    #[unstable(feature = "iter_exhaust", issue = "44546")]
+    fn exhaust(self) where Self: Sized {
+        for _ in self {}
```
Comment (cuviper) [inline src/libcore/iter/iterator.rs]: This should use `for_each` to get internal iteration, rather than (implicit) `next()` calls.

Comment (cuviper) [reply src/libcore/iter/iterator.rs]: The difference is that some iterators can apply a `fold` much more efficiently than they can `next` one item at a time.  For instance, `Chain::fold` can fold its first part as a whole, then its second part as a whole, but `Chain::next` has to check whether it's on the first or second part every time.  This is demonstrated in the benchmarks added to the `for_each` PR, #42782, in commit 4a8ddac99e1edfb219e11c3ea2d6c43ccecb29ab.

**Pull request closed without merging**